analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs

```diff
@@ -0,0 +1,42 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Helpers;
+
+public enum TrustedCodeSource
+{
+    Roslyn,
+    StyleCop
+}
+
+[AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct | AttributeTargets.Interface | AttributeTargets.Method | AttributeTargets.Constructor | AttributeTargets.Property,
+    AllowMultiple = true, Inherited = false)]
+public sealed class CodeCopiedFromAttribute : Attribute
+{
+    public CodeCopiedFromAttribute(TrustedCodeSource trustedCodeSource, string permanentUrl)
+    {
+        TrustedCodeSource = trustedCodeSource;
+        PermanentUrl = permanentUrl;
+    }
+
+    public TrustedCodeSource TrustedCodeSource { get; }
+    public string PermanentUrl { get; }
+    public string Reasoning {  get; set; }
```

Comment:
Any reason to have `Reasoning` a read-write property? 
I would make readonly, as all the other ones, and use a primary constructor:
```cs
[AttributeUsage(AttributeTargets.All, AllowMultiple = false, Inherited = false)]
public sealed class CodeCopiedFromAttribute(
    TrustedCodeSource TrustedCodeSource, string PermanentUrl, string Reasoning) : Attribute;
```

Reply:
Because it is optional.

Reply:
Default parameters for attributes are considered bad practise because of metadata bloat. It is only a single byte more, but still:
```
    .custom instance void MyAttribute::.ctor(string) = (
        01 00 ff 00 00
    )
    .custom instance void My2Attribute::.ctor() = (
        01 00 00 00
```
https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgDaBZAngCgJQC6qmWc+RKMAzAATw0DCNA3qje3bTACw0b4s2HAL6pRKVNTpw+WAILBgAJzAAjAK7AApjRA0FytZq2pWKDp1kGVG7TgDKhgHYBzGtoDOwGgF4aT9QgIPCF2MwtxcUkuGWw4ayNtXX1FG2NTVFDLGAQABhoAFS0vFhoXLWAAbhoPCurIoA=

Reply:
Alright, 1 byte doesn't seem enough to me to talk about "metadata bload", but fair enough.